sdk/src/main/java/io/opentelemetry/sdk/trace/SpanBuilderSdk.java

```diff
@@ -223,13 +223,7 @@ public Span startSpan() {
         traceConfig
             .getSampler()
             .shouldSample(
-                parentContext,
-                traceId,
-                spanId,
-                spanName,
-                spanKind,
-                immutableAttributes,
-                immutableLinks);
+                parentContext, traceId, spanName, spanKind, immutableAttributes, immutableLinks);
```

Comment:
One of the goals of this change in the spec was to enable deferring the generation of a span id when one wasn't needed. Is it possible to defer the spanId creation, and if not, why not?

Reply:
@jkwatson To be clear, you are talking about `SpanId` being crated unconditionally before this line? e.g.

```
SpanId spanId = idsGenerator.generateSpanId();
```

?

Reply:
@carlosalberto Exactly, this was mentioned on the original spec PR (https://github.com/open-telemetry/opentelemetry-specification/pull/621) to improve performance.

Reply:
Follow-up idea (for spec): why not also generate trace ID later? And provide an option for samplers that need it (only one single one: ProbabilitySampler) to optionally generate a trace ID themselves that the SDK will use.

Reply:
Yeah, I was focusing on avoiding samplers starting to use the `SpanId` and missed the optimization. Let me see what can be done.

Reply:
Changed the code to use `SpanId.getInvalid()` if the span is not sampled and only generate a new `SpanId` if it is sampled.

Reply:
Please revert the last commit, and don't do the optimization now.

Reply:
Done, reverted the optimization - didn't see your comment on GH prior to the change...